### DISCUSSION TITLE
writer object

### DISCUSSION THREAD

**Discussion by Noiders** [weltenwort/home-assistant-rct-power-integration#277] [Ideas]
Hi,

I've been experimenting a bit with rctclient repo and found a way to somewhat control the inverter so I can better control the grid input/output and battery charge/discharge. Problem is I am a complete novice at programming Python and looking at the code you wrote to integrate it with Homeassistant I get completely lost.

As of now, there's only the reader part. I was wondering if we could add 2 enteties for writing. Followin entities from the rctclient.registry are writable: 

> ObjectInfo(group=ObjectGroup.POWER_MNG,       
> object_id=0xBD008E29, 
> index=667, 
> request_data_type=DataType.FLOAT,  
> unit='W',           
> name='power_mng.battery_power_extern',               
> description='Battery target power (positive = discharge)')

with this you can externaly control the battery charge/discharge. And by doing calculations, you can control gridload.

> ObjectInfo(group=ObjectGroup.POWER_MNG,
> object_id=0xF168B748, 
> index=835, 
> request_data_type=DataType.ENUM,                       
> name='power_mng.soc_strategy',                       
> description='SOC target selection',
> enum_map={0: 'SOC target = SOC', 1: 'Constant', 2: 'External', 3: 'Middle battery voltage', 4: 'Internal', 5: 'Schedule'}),

with this, you can set the SOC control to 2 (External) so that the previous entity can control the battery charge.

thanks for looking into this

N.

**Comment by weltenwort** [comment]
Hey @Noiders, thanks for reaching out!

I've shied away from implementing write access so far because it has the potential to damage the battery. Given how "bare-bones" the inverter firmware seems to be I don't have much confidence that it would appropriately guard against dangerous settings.

I'd consider it more readily if there was official, public documentation for the inverter's API. Without it I just don't have the confidence that this integration would remain harmless to the average user.

Do you have any idea how to resolve that tension?

**Comment by Noiders** [comment]
Hi @weltenwort, 

I've been doing some (messy) programming in Python and have been able to control the battery with these 2 settings. All i can add is: this works and did not lead to any issues. I've even send out of bound values for battery charging and the inverter corrected it itselve. So on that part I don't expect any issue.

I've also have been in contact with RCT, but as you probably know, they refuse to deliver any support.
I've also asked if they were planning to implement settings to:
- configure the plant to sustain a same maximum load on the grid by using the battery as backup (more on this later).
- configure the plant to maintain a battery soc percentage in order to sustain the first option.
but they wont have it.

In Belgium we are as of 1/1/2023 charged on how much we stress the grid. Each month Fluvius (the company that monitors the grid) writes down your peak load over 15 mintues. In the end of the year an average over 12 months is taken and you'll have to pay a tax for that. So, beïng able to limit the maximum load on the grid and use the battery as a buffer, the lower the tax will be. Frankly this is an option allready available in multiple brands and is refered to as Peak Shaving. RCT is not offering this option and will not program it.
the small program i wrote is capable of doing this peak shaving by doing a simple calculation of the current grid load and telling the battery to load or discharge accordingly.

Next to that, I would also want make use of hourly pricing for energy. Every day right before 14:00 (2PM) the prices of electricity for the next 24hours is posted. I would then have my Battery (especialy in winter) charge full during the night, when prices are low. I allready have this info in my Home Assistant. I also can control the charging of my EV using Home Assistant. So having all this info allready in Home Assistant it would make it alot easyer to link the info into actions using NodeRed.

So, unfortunately, I cannot resolve that tension. All I can offer is a live testing environment. And sharing of knowledge once I have the configuration for Peak Shaving and (lets call it ) Price shaving up and running.

Most probably we would be able to limit the variables we can write back to the inverter and that way make sure we only publish what we have tested to work?

thanks for considering.

**Comment by weltenwort** [comment]
Thanks for explaining the use case. It makes a lot of sense and I'd love to support energy-efficiency initiatives like that.

I like that your approach uses the external producer feature, which by design should be built to interface with potentially unreliable third-party devices. It doesn't sound as risky as I initially thought as long as we are selective about which write entities we expose. And we could mark them as being disabled by default in Home Assistant so it's less likely that someone will tinker with them unintentionally.

Now there's only the matter of finding the time to implement it. I don't have a lot of spare time currently, but that doesn't mean I wouldn't pick it up in the future.

If I were to implement it I would probably expose `power_mng.soc_strategy` as an [MQTT select](https://www.home-assistant.io/integrations/select.mqtt/) entity and `power_mng.battery_power_extern` as an [MQTT number](https://www.home-assistant.io/integrations/number.mqtt/) entity.

That should allow for usage in an automation as you describe it, right?

**Reply by Noiders** [reply]
I suppose so, yes. Sorry but I'm no programmer. I have some knowledge, but looking at the knowledge needed for programming home assistant plugins is way over my head. I'd love to learn though.

**Reply by weltenwort** [reply]
Oh, sorry, I didn't mean to put any pressure on you. Just wanted to check that I understood your use-case correctly.

**Reply by Noiders** [reply]
no need to apologize. I know my limits.

**Reply by raserj** [reply]
can I ask if you have moved somewhere? I would also like to set the charging mode according to spot prices

**Reply by weltenwort** [reply]
Unfortunately not, sorry. But this is still a valid request that nobody has gotten around to implementing yet. Thanks for indicating more demand, though.

**Comment by HWE70** [comment]
@Noiders In fhempy the implementation was done in November. power_mng.battery_power_extern and power_mng.soc_strategy can be set accordingly, and when I read the values afterwards they are set as expected e.g. "2" for power_mng.soc_strategy and "-1000.00" for power_mng.battery_power_extern to force loading the battery with 1000W. However, for me it does not work. Yes, the battery starts charging even though the power_mng.soc_min (0.70) was already reached, that's fine, but it is not charging from grid. It only charges the energy currently available from the PV panels. Is there something else to do? Can you advise your firmware versions (inverter and bms)?

**Comment by HWE70** [comment]
Hi! It works now. Of course one need to enable "external energy" in general, to make it work. Furthermore the battery SOC level needs to be above SOC min. Otherwise external charging is limited to **power_mng.soc_charge_power**, wich is 100W by default. Above SOC min  **power_mng.battery_power_extern** is used as expected :-)

**Comment by Gizmo0815** [comment]
It would be very helpful to be able to set batt_soc_min via the integration. I just keep my fhempy running because of the soc management I have implemented there.

**Comment by do-gooder** [comment]
https://github.com/do-gooder/rctpower_writesupport

**Reply by raserj** [reply]
if I set the power_mng.soc strategy to 2, my home assistant crashes, does it happen to you too? I don't have the External option in the application, so maybe that's why?

**Reply by heikone** [reply]
Same for me! HA crashes!
But if I'm running the script outside Homeassistant by command ```rct.py set power_mng.soc_strategy 2 --host=your_IP_here``` it works. Even if the option 'External' is not provided as selection, the app reports 'External' after setting by command.
Also testet to charge and discharge by a certain power works by ```rct.py set power_mng.battery_power_extern -250 --host=your_IP_here```.
@raserj: Maybe you can try to set strategy outside HA first and report if it works?
@do-gooder: Any idea, why HA crashes in this case while it works with setting 'soc_max' as I proposed here [Homeassistant Integration with pyscript](https://github.com/do-gooder/rctpower_writesupport?tab=readme-ov-file#implementation-as-service)?

**Reply by do-gooder** [reply]
The “External” option has been removed for customers and installers from the app. However, it can be set via the rctclient.

I have no idea why this does not work via pyscript. For my purposes,  [Home Assistant Integration as package](https://github.com/do-gooder/rctpower_writesupport?tab=readme-ov-file#home-assistant-integration-as-package) is sufficient.

**Reply by heikone** [reply]
Solution found!

Try out this change in `rct_ha_call.py`:
https://github.com/do-gooder/rctpower_writesupport/commit/ab6f9f724e8f520004ce8ee443e817383cb1abac

Greetings, Heiko

**Reply by do-gooder** [reply]
@raserj: Did the patch solve your problem?

**Reply by raserj** [reply]
@heikone Thank you for the help, but I resolved it by setting the Schedule to a specific power output in the app, and I simply use Home Assistant to switch between Constant and Schedule. This is enough for me to charge the battery at a low price.

**Reply by heikone** [reply]
@raserj: Good and practical idea! Thanks for your advice to use schedule.
Maybe, my new issue in repo rtcpower_writesupport ([link](https://github.com/do-gooder/rctpower_writesupport/issues/9#issue-2531831315)) is an alternative to schedules.

**Reply by Ovik10** [reply]
Yo guys, Im searching for the solution for charging the batteries via grid. Can you explain me a little bit more how to integrate it into the HA? Thanks a lot.

**Reply by heikone** [reply]
Did you check out our repository [rctpower_writesupport](https://github.com/do-gooder/rctpower_writesupport.git)?
Your question, @Ovik10, sounds like the use case [here](https://github.com/do-gooder/rctpower_writesupport?tab=readme-ov-file#charging-from-the-grid-for-eg-tibber) and the relevant implementation example of @do-gooder [here](https://github.com/do-gooder/rctpower_writesupport?tab=readme-ov-file#home-assistant-integration-as-package).
Just let us know, if you a have a specific question!
Heiko

**Reply by DrZoidberg89** [reply]
Hi @heikone, 
I am also very interested in your project and have been following it for a while. But I still don't have the feeling whether this is currently foolproof "idiotensicher"? I think Ovik10 would like to have more screenshots of the application and how exactly to install everything, this would also help me to approach your project.

Is it possible that you will create a HACS for this in the future, which is very easy to install. Or even an integration into the RCT Power Integration of Weltenwort?

I would also be willing to make a financial contribution if it would be more user-friendly for non-programmers. :)

**Reply by heikone** [reply]
Dear DrZoidberg89,
I understand that the solution do-gooder and I created is not fool-proof!
When we started with the repo, the intention was simply to fulfill our requests.
Since do-gooder and I might  have more experience with Home Assistant and are used "programmer's solutions", our repo is sufficient for us.
Creating a HACS solution would mean to spend time for planning, setting up, maintenance, ... and at least I'm lacking of time beside my job.
**Anyway, to make things work for you, I would propose, that you start a discussion or/and issue in our [repo](https://github.com/do-gooder/rctpower_writesupport.git) and explain, were you stuck in our installation instruction. I would appreciate to receive your questions and help you!**
We can than use your questions to update and improve our instructions.
This would be a faster way for you, instead of waiting for a HACS solution.
Greetings, Heiko
